bazel/PPROF.md

```diff
@@ -155,3 +175,145 @@ More complex flame/graph charts can be generated and viewed in a browser, which
+    -U "/full/path/to/envoy/bazel-bin/source/exe/envoy-static:_Znwm" > out.stack
+$ ./FlameGraph/stackcollapse.pl < out.stacks | ./FlameGraph/flamegraph.pl --color=mem \
+    --title="operator new(std::size_t) Flame Graph" --countname="calls" > out.svg
+```
+
+The `stackcount` utility counts function calls and their stack traces using eBPF probes.
+Since Envoy by default links statically to tcmalloc which provides its own implementation
+of memory management functions the used uprobe looks like
+```/full/path/to/envoy/bazel-bin/source/exe/envoy-static:_Znwm```. The part before
+the colon is a library name (a path to Envoy's binary in our case). The part after the
+colon is a function name as it looks like in the output of `objdump -tT /path/to/lib`,
+that is mangled in our case. To get an idea how your compiler mangles the name you
+can use this one-liner:
+```
+$ echo -e "#include <new>\n void* operator new(std::size_t) {} " | g++ -x c++ -S - -o- 2> /dev/null
+        .file   ""
+        .text
+        .globl  _Znwm
+        .type   _Znwm, @function
+_Znwm:
+.LFB73:
+        .cfi_startproc
+        pushq   %rbp
+        .cfi_def_cfa_offset 16
+        .cfi_offset 6, -16
+        movq    %rsp, %rbp
+        .cfi_def_cfa_register 6
+        movq    %rdi, -8(%rbp)
+        nop
+        popq    %rbp
+        .cfi_def_cfa 7, 8
+        ret
+        .cfi_endproc
+.LFE73:
+        .size   _Znwm, .-_Znwm
+        .ident  "GCC: (GNU) 10.2.1 20201016 (Red Hat 10.2.1-6)"
+        .section        .note.GNU-stack,"",@progbits
+```
+
+WARNING: The name is going to be different on 32-bit and 64-bit platforms due to different sizes
+of `size_t`. Also ```void* operator new[](std::size_t)``` is a separate function as well as `malloc()`.
+The latter is a C function and hence not mangled by the way.
+
+`stackcount` doesn't count how much memory is allocated, but how often. To answer the "how much"
+question you could use Brendan's
+[mallocstacks](https://github.com/brendangregg/BPF-tools/blob/master/old/2017-12-23/mallocstacks.py)
+tool, but it works only for `malloc()` calls. You need to modify it to take into
+account other memory allocating functions.
+
+# Performance annotations
+
+In case there is a need to measure how long a code path takes time to execute in Envoy you may
+resort to instrumenting the code with the
+[performance annotations](https://github.com/envoyproxy/envoy/blob/master/source/common/common/perf_annotation.h).
+
+There are two types of the annotations. The first one is used to measure operations limited by
+a common lexical scope. For example:
+
+```c++
+void doHeavyLifting() {
+  PERF_OPERATION(op);
+  bool success = doSomething();
+  if (success) {
+    finalizeSuccessfulOperation();
+    PERF_RECORD(op, "successful", "heavy lifting");
+  } else {
+    recoverFromFailure();
+    PERF_RECORD(op, "failed", "heavy lifting")
+  }
+}
+```
+
+The recorded performance data can be dumped to stdout with a call to `PERF_DUMP()`:
```

Comment:
As an aside, it would be pretty neat if somehow on builds that have this compiled in there was an admin endpoint that could be used to dump ongoing traces? cc @jmarantz